src/util/db/dbid.h

```diff
@@ -1,126 +1,166 @@
+/// Base class for ID values of objects that are stored in the database.
 //
-// Implicit conversion from/to the native value type has been disabled
-// on purpose in favor of explicit conversion. The internal value type
-// is also hidden as an implementation detail.
+/// Implicit conversion from/to the native value type has been disabled
+/// on purpose in favor of explicit conversion. The internal value type
+/// is also hidden as an implementation detail.
 //
-// Although used as a base class this class has a non-virtual destructor,
-// because destruction is trivial and for maximum efficiency! Derived
-// classes may add additional behavior (= member functions), but should
-// not add any additional state (= member variables). Inheritance is
-// only needed for type-safety.
+/// Although used as a base class this class has a non-virtual destructor,
+/// because destruction is trivial and for maximum efficiency! Derived
+/// classes may add additional behavior (= member functions), but should
+/// not add any additional state (= member variables). Inheritance is
+/// only needed for type-safety.
 class DbId {
-protected:
-public:
-    // Alias for the corresponding native type. It keeps the
-    // implementation of this class flexible if we ever gonna
-    // need to change it from 'int' to 'long' or any other type.
+  public:
+    /// Alias for the corresponding native type. It keeps the
+    /// implementation of this class flexible if we ever gonna
+    /// need to change it from 'int' to 'long' or any other type.
     typedef int value_type;
 
-    DbId()
-        : m_value(kInvalidValue) {
-        DEBUG_ASSERT(!isValid());
+    static constexpr value_type kInvalidValue = -1;
+    static constexpr value_type kMinValue = 0;
+
+    static constexpr bool isValidValue(value_type value) {
+        static_assert(kInvalidValue < kMinValue);
+        return value >= kMinValue;
     }
-    explicit DbId(value_type value)
-        : m_value(value) {
-        DEBUG_ASSERT(isValid() || (kInvalidValue == m_value));
+
+
+    constexpr DbId()
+            : m_value(kInvalidValue) {
     }
     explicit DbId(QVariant variant)
-        : DbId(valueOf(std::move(variant))) {
+            : DbId(NoAssert{}, valueOf(std::move(variant))) {
     }
 
-    bool isValid() const {
+    /// Explicit checked conversion from inner value
+    ///
+    /// Creates an instance from the provided inner value. The value
+    /// might be either valid or invalid. Triggers a debug assertion
+    /// if an invalid value doesn't equal the predefined constant.
+    explicit DbId(value_type value)
+            : DbId(NoAssert{}, value) {
+        DEBUG_ASSERT(isValid() || m_value == kInvalidValue);
+    }
+
+    /// Explicit unchecked conversion from inner value
+    ///
+    /// Creates an instance with the provided inner value without
+    /// any consistency checks that could trigger an assertion.
+    static constexpr DbId fromValueUnchecked(value_type value) {
```

Comment:
This is unused and can be removed. 
Than DbId(NoAssert{}, value); can be removed along with it.

Reply:
It is used in tests.

Reply:
For what? If I get this right it test condition that can now no longer happen due to the assertions in the constructor. 
If you plan to use this feature in productive code, please explain the use case. 
If not please remove it to not cutter this class with useless test fixtures.

Reply:
The debug assertion doesn't prevent this from happening in release builds. When initialized with an unexpected value conversion back into a QVariant or QString should still handle even this case correctly. I will not delete those tests.

Reply:
Ah OK, in this case, it would be much better to replace the undefined value with kInvalidValue in the constructor. 
And the case you describe will never happen.